README.rst
nbodykit: a massively parallel large-scale structure toolkit
============================================================

.. image:: https://zenodo.org/badge/34348490.svg
   :target: https://zenodo.org/badge/latestdoi/34348490

|

**nbodykit** is an open source project and Python package providing
a set of algorithms useful in the analysis of cosmological
datasets from N-body simulations and large-scale structure surveys.

Driven by the optimism regarding the abundance and availability of
large-scale computing resources in the future, the development of nbodykit
distinguishes itself from other similar software packages
(i.e., `nbodyshop`_, `pynbody`_, `yt`_, `xi`_) by focusing on :

- a **unified** treatment of simulation and observational datasets by
  insulating algorithms from data containers

- reducing wall-clock time by **scaling** to thousands of cores

- **deployment** and availability on large, super-computing facilities

- an **interactive** user interface that performs as well in a `Jupyter
  notebook <http://jupyter.org>`_ as on super-computing machines

All algorithms are parallel and run with Message Passing Interface (MPI).

.. _nbodyshop: http://www-hpcc.astro.washington.edu/tools/tools.html
.. _pynbody: https://github.com/pynbody/pynbody
.. _yt: http://yt-project.org/
.. _xi: http://github.com/bareid/xi
.. _`NERSC`: http://www.nersc.gov/systems/

Build Status
------------

We perform integrated tests of the code, including all built-in algorithms, in a
miniconda environment for Python 2.7, 3.5, and 3.6.

.. image:: https://travis-ci.org/bccp/nbodykit.svg?branch=master
    :alt: Build Status
    :target: https://travis-ci.org/bccp/nbodykit
.. image:: https://coveralls.io/repos/github/bccp/nbodykit/badge.svg?branch=master
    :alt: Test Coverage
    :target: https://coveralls.io/github/bccp/nbodykit?branch=master
.. image:: https://img.shields.io/conda/v/bccp/nbodykit.svg
   :alt: Conda
   :target: https://anaconda.org/bccp/nbodykit
.. image:: https://img.shields.io/pypi/v/nbodykit.svg
   :alt: PyPi
   :target: https://pypi.python.org/pypi/nbodykit/

Documentation
-------------

The official documentation is hosted on ReadTheDocs at http://nbodykit.readthedocs.org/.

Cookbook Recipes
----------------

Users can dive right into an interactive cookbook of example recipes using binder.
We've compiled a set of Jupyter notebooks to help users learn nbodykit by example — just click the launch button below to get started!

.. image:: http://mybinder.org/badge.svg
    :alt: binder
    :target: https://mybinder.org/v2/gh/bccp/nbodykit-cookbook/master?filepath=recipes

|

Users can also view a static version of the cookbook recipes
`in the documentation <http://nbodykit.rtfd.io/en/latest/cookbook/index.html>`_.

Installation
------------

We recommend using the Anaconda distribution of Python. To obtain the
dependencies and install a package on OSX or Linux, use

.. code-block:: bash

    $ conda install -c bccp nbodykit

We are considering support for Windows, but this depends on the status
of `mpi4py`.

Using nbodykit on NERSC
-----------------------

On the Cori and Edison machines at NERSC, we maintain a nightly conda build of
the latest stable release of nbodykit. See
`the documentation <http://nbodykit.readthedocs.io/en/latest/install.html#nbodykit-on-nersc>`_
for using nbodykit on NERSC for more details.

Bumping to a new version
------------------------

1. git pull - confirm that the master branch is up-to-date
2. Edit Changelog (CHANGES.rst) - Make sure to include all issues which have arisen since the last version. (git add ... -> git commit -m "Update Changelog" -> git push)
3. Edit version.py -> git push ("bump version to ...")
4. Go to https://travis-ci.org/bccp/nbodykit and make sure it merged without any problems.
5. Go to bccp/conda-channel-bccp repo and do "Restart build"
6. git tag 0.3.? -> git push --tags
7. bump to a development version (0.3.?dev0)


Acknowledgement
---------------

The work on nbodykit is supported by
the Berkeley Center for Cosmological Physics, University of California Berkeley,
by
the National Energy Research Scientific Computing Center of the Lawrence Berkeley National Lab via the allocation m3035,
and by users and contributors.


CHANGES.rst
Changelog
=========

0.3.16 (UNRELEASED)
------------------

0.3.15 (2020-09-08)
------------------
* :issue:`628`: FileType and FileCatalog clean up.
* :issue:`630`: Fix wrong shape (length of 6 instead of 3) of complex mesh stored in attrs.
* :issue:`640`: update for pmesh 0.1.56
* :issue:`641`: update for pandas 1.1.

0.3.14 (2020-07-08)
-------------------
* :issue:`622`: meshtools sets mu to zero instead of nan when appropriate.
* :issue:`618`: zhist fix
* :issue:`613`: gaussian filter fix (was not applied)
* :issue:`610`: odd multipole and wide angle fix in ConvFFTPower
* :issue:`604`: Allow any glob pattern.

0.3.13 (2019-08-01)
-------------------
* :issue:`000`: Update for new dask (csv)
* :issue:`597`: Update docrep.py
* :issue:`596`: Update documentation for NERSC usage. Python 3.5 is no longer supported on NERSC.

0.3.12 (2019-07-06)
-------------------
* :issue:`592`: Update for new dask (ndim)
* :issue:`587`: Reserve label == 0 for unconnected particles.
* :issue:`586`: Center the particles in fiber collision
* :issue:`582`: documentation infra update.

0.3.11 (2019-04-28)
------------------
* :issue:`575,576`: Fix bug in painting particles to mesh in CatalogMesh
* :issue:`569`: Upgrade DistributedArray
* :issue:`570`: Add initial Query interface to FileCatalog.
* :issue:`571`: Typo in FKP
* :issue:`573`: Support Gadget variant binary format (with 4 byte header)
* :issue:`574`: DistributedArray.bincount fix
* :issue:`577`: Add doc examples for CVS and ArrayCatalog.
* :issue:`579`: compat-fix for mcfit-0.0.16
* :issue:`581`: compat-fix for sympy-1.4

0.3.10 (2019-02-07)
------------------
* :issue:`555,556`: Impove dask interpolation in Catalog.save.
* :issue:`557`: Use dask's gufunc for sky transforms
* :issue:`558`: Note the unit of HaloRadius (proper, not comoving. bite me)
* :issue:`561`: Allow setting the position column in 2PCF.
* :issue:`563`: RedshiftHistogram extrapolates to zero rather than any number.
* :issue:`564`: Fix missing compensations with Multi-species meshes.
* :issue:`565`: add keyward header to HDFCatalog for reading additional meta data
* :issue:`566`: fix a bug sorting on float32, and add a persist method to Catalog.
* :issue:`567`: suppress redundant output in convpower.

0.3.9 (2019-01-07)
------------------
* :issue:`544,549`: Use dask.store to save a catalog.
* :issue:`545`: Fix shotnoise estimation of weighted to_mesh() calls.
* :issue:`546`: Fix IndexError in painting during throttling.
* :issue:`547`: Update docrep
* :issue:`548`: Fix many deprecation warnings
* :issue:`550`: Documentation updates
* :issue:`553`: Fix error in TPCF module when there is only 1 bin.

0.3.8 (2018-12-29)
------------------
* :issue:`543`:  Further performance improvements on catalog slicing.
* :issue:`542`:  The IO module shall make sure buffer is c-contiguous before reshaping
* :issue:`541`:  Allow setting cartesian / sphericial transformation reference frame
* :issue:`540`:  Allow not saving the header in Catalog.save
* :issue:`539`:  Allow non-uniform redshifts in halo property transformations.
* :issue:`538`:  Stop gathering catalog to a single rank in HaloCatalog
* :issue:`537`:  Use numpy.sum for summing of integers.
* :issue:`536`:  Fix boxsize mismatch comparision in pair counters.
* :issue:`535`:  Improve working with a dask cluster.
* :issue:`532`:  Improve speed of slicing of a catalog.
* :issue:`531`:  Additional throttling during painting.
* :issue:`530`:  Use setuptools (need to change conda-build-bccp recipe)
* :issue:`529`:  Add kmax(rmax) to FFTPower, FFTCorr, ConvPower.
* :issue:`528`:  Add dataset= to Catalog.save, deprecate datasets=[]

0.3.7 (2018-10-17)
------------------
* :issue:`519`:  Rework the class hierarchy of Catalogmesh.
* :issue:`526`:  Reduce the paint size for systems with lower mem per core
* :issue:`527`:  Aggregate attrs of header and the main datasets.

0.3.6 (2018-09-26)
------------------
* :issue:`518`:  Rework CurrentMPIComm
* :issue:`521`:  Fix OOM errors with dask >= 0.19.0

0.3.5 (2018-08-23)
------------------
* :issue:`509`:  Fix auto detection of f8 type in Gadget1 file reader
* :issue:`513`:  Ignore divide errors.
* :issue:`516`:  Fix several bugs in three point function
* :issue:`517`:  Improve compatibility with numpy 1.15.x's new indexing convention.

0.3.4 (2018-06-29)
------------------
* :issue:`495`:  Improve scaling of LogNormal catalog
* :issue:`497`:  take method to BinnedStatistic
* :issue:`498`:  add compute method to Catalog interface; CatalogMesh no longer a Catalog
* :issue:`500`:  unique binning in FFTPower and FFTCorr
* :issue:`503`:  redistributing a catalog spatially
* :issue:`504`:  Catalog.copy hangs
* :issue:`505`:  update docrep to 0.2.3
* :issue:`506`:  compatible with dask 0.18.1.

0.3.3 (2018-05-30)
------------------
* :issue:`491`:  update compatibility with pandas 0.23.0 in cgm.
* :issue:`490`:  write more useful weights and pairs in the paircount result.
* :issue:`493,494`:  update for deprecation in pmesh

0.3.2 (2018-05-14)
------------------
* :issue:`475`:  proper normalization of the Landy-Szalay estimator, included R1R2 option and to_xil function
* :issue:`487`:  Linear theory correspondant of nbody simulation. (three fluid model)
* :issue:`486`:  overdecomposition in FOF
* :issue:`483`:  switching to a new type in BinnedStatistics.copy()
* :issue:`482`:  Fix a crash when two datasets passed into corrfunc are of different dtypes.
* :issue:`480`:  BigFileCatalog shall look for header relative to the root of file.
* :issue:`479`:  GatherArray allows root=Ellipsis (for allbather)
* :issue:`476`:  Fix MeshSource.apply if MeshSource.action is overriden
* :issue:`471`:  Decompose of surveydata to the correct bounds.

0.3.1 (2018-04-10)
------------------
* :issue:`468`:  corrfunc and big-endian floating point numbers
* :issue:`470`:  Add hankel tranforms for ell>0 
* :issue:`469`:  Fix a regression painting 'apply'ed meshes.

0.3.0 (2017-12-18)
------------------

benchmark.py
from __future__ import print_function
import argparse
import os
from jinja2 import Environment, FileSystemLoader, Template
import tempfile
import subprocess
import numpy

# valid benchmarking sample names
SAMPLES = ['boss_like', 'desi_like', 'dm_like']

# the directory this file lives in
toplevel = os.path.split(os.path.abspath(__file__))[0]

def minutes_to_job_time(minutes):
    h, m = divmod(minutes, 60)
    return "%02d:%02d:00" % (h, m)

def get_nodes_from_cores(cores, host):
    if host == 'cori':
        nodes, extra = divmod(cores, 32)
    elif host == 'edison':
        nodes, extra = divmod(cores, 24)
    else:
        raise ValueError("bad host name '%s'" %host)
    # account for remainder cores
    if extra > 0: nodes += 1
    return nodes

class NERSCBenchmark(object):
    """
    Class to submit benchmark tasks to NERSC.
    """

    @staticmethod
    def load(path):
        """
        Return a NERSC benchmark result as a pandas dataframe.

        Parameters
        ----------
        path : str
            the path to a directory holding the results; this directory
            should hold a ``config.json`` file.
        """
        import pandas as pd
        import json
        
        # make sure config file exists first
        cfg_file = os.path.join(path, 'config.json')
        if not os.path.exists(cfg_file):
            raise ValueError("the path should contain a 'config.json' file")

        # load the config so we can attach to dataframe
        config = json.load(open(cfg_file, 'r'))

        # walk the directory and load .json files
        rows = []
        for dirpath, dirnames, filenames in os.walk(path):

            # check all files for .json extension
            for filename in filenames:

                if filename != 'config.json' and filename.endswith('.json'):

                    # first path is sample, if it exists
                    sample = os.path.relpath(os.path.normpath(dirpath), path).split(os.path.sep)[0]
                    if sample not in SAMPLES:
                        sample = ""

                    # load the JSON file
                    d = json.load(open(os.path.join(dirpath, filename), 'r'))

                    # add a row for each tagged benchmark in each test function
                    for test in d['tests']:
                        for tag in d['tags']:
                            x = d[test][tag]
                            row = (sample, test, tag, len(x))
                            row += (numpy.median(x), numpy.mean(x), numpy.min(x), numpy.max(x))
                            rows.append(row)

        # make the data frame
        names = ['sample', 'name', 'tag', 'ncores', 'dt_median', 'dt_mean', 'dt_min', 'dt_max']
        df = pd.DataFrame(data=rows, columns=names)
        df.config = config
        return df

    @classmethod
    def get_parser(cls):

        desc = 'run the specified benchmark on NERSC'
        parser = argparse.ArgumentParser(description=desc)

        h = 'the path of the benchmark to run in the nbodykit source code'
        parser.add_argument('benchname', type=str, help=h)

        h = 'the name of the sample to run'
        parser.add_argument('--sample', type=str, default=None, choices=SAMPLES, help=h)

        h = 'the output path to save the benchmark result to'
        parser.add_argument('--bench-dir', type=str, help=h, required=True)

        h = 'the Python version to use'
        choices = ['2.7', '3.5', '3.6']
        parser.add_argument('--py', type=str, choices=choices, default='3.6', help=h)

        h = 'the number of nodes to request'
        parser.add_argument('-n', '--cores', type=int, help=h, required=True)

        h = 'the NERSC partition to submit to'
        choices=['debug', 'regular']
        parser.add_argument('-p', '--partition', type=str, choices=choices, default='debug', help=h)

        h = 'the requested amount of time (in minutes)'
        parser.add_argument('-t', '--time', type=int, default=30, help=h)

        h = 'the nbodykit source tag to checkout before running'
        parser.add_argument('--tag', type=str, default='master', help=h)

        return parser

    @classmethod
    def submit(cls):
        """
        Submit the benchmarking job on NERSC.
        """
        parser = cls.get_parser()
        ns = parser.parse_args()
        cls._submit(ns)

    @staticmethod
    def _submit(ns):
        """
        Internal function responsible for submiting the task, using the
        parsed command-line Namespace ``ns``.
        """
        # determine the NERSC host
        host = os.environ.get('NERSC_HOST', None)
        if host is None:
             raise RuntimeError("benchmark.py should be executed on NERSC")

        # make the benchmark directory absolute
        ns.bench_dir = os.path.normpath(os.path.abspath(ns.bench_dir))

        #  compute the job name and job script output path
        output_dir = os.path.join(ns.bench_dir, 'output')
        if not os.path.exists(output_dir):
            os.makedirs(output_dir)
        output_file = "{host}-%j.out.{cores}".format(host=host, cores=ns.cores)
        output_file = os.path.join(output_dir, output_file)

        # load the template job script
        jinja_env = Environment(loader=FileSystemLoader(toplevel))
        tpl = jinja_env.get_template('job.template.sh')

        # the configuration to pass to the template
        config = {}
        config['benchname'] = ns.benchname
        config['benchdir'] = ns.bench_dir
        config['partition'] = ns.partition
        config['time'] = minutes_to_job_time(ns.time)
        config['python_version'] = ns.py
        config['tag'] = ns.tag
        config['cores'] = ns.cores
        config['nodes'] = get_nodes_from_cores(ns.cores, host)
        config['haswell_config'] = "#SBATCH -C haswell" if host == 'cori' else ""
        config['output_file'] = output_file
        config['job'] = "benchmark.py"

        # the option to select specific sample
        if ns.sample is not None:
            config['sample'] = "-m %s" %ns.sample
        else:
            config['sample'] = ""

        # render the template
        rendered = tpl.render(**config)

        # write to temp file and call
        with tempfile.NamedTemporaryFile(mode='w') as ff:

            # echo the job scripts
            print(rendered)

            # write to temp file (and rewind)
            ff.write(rendered)
            ff.seek(0)

            # and call
            subprocess.call(["sbatch", ff.name])

if __name__ == '__main__':
    NERSCBenchmark.submit()
